corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -615,26 +616,27 @@ def save_new_form(form):
         """
         Save a previously unsaved form
         """
+        from corehq.form_processor.backends.sql.processor import get_tracked_objects_db_names
         if form.is_saved():
             raise XFormSaveError('form already saved')
         logging.debug('Saving new form: %s', form)
 
-        operations = form.get_tracked_models_to_create(XFormOperationSQL)
-        for operation in operations:
-            if operation.is_saved():
-                raise XFormSaveError(
-                    'XFormOperationSQL {} has already been saved'.format(operation.id)
-                )
-            operation.form_id = form.form_id
-
+        db_names = {form.db}
+        db_names.update(get_tracked_objects_db_names(form, raise_error_if_saved=True))
         try:
             with form.attachment_writer() as attachment_writer, \
-                    transaction.atomic(using=form.db, savepoint=False):
+                    ExitStack() as stack:
+                for db_name in db_names:
+                    stack.enter_context(
+                        transaction.atomic(db_name, savepoint=False))
                 transaction.on_commit(attachment_writer.commit, using=form.db)
                 form.save()
                 attachment_writer.write()
-                for operation in operations:
-                    operation.save()
+                for model_class, model_objects in form.create_models.items():
+                    for model_object in model_objects:
+                        if hasattr(model_object, 'form_id'):
+                            model_object.form_id = form.form_id
+                        model_object.save()
```

Comment:
on slack you mentioned saving the vault entries separately, if you do that then these changes are not necessary

Reply:
Ya that was the implementation i was considering earlier. 
This is where I changed it in https://github.com/dimagi/commcare-icds/pull/41/commits/bf622c0d16eac60734cdd4f0cbefbe5780ce09a1.
There are definite advantages if we hook into hq like
1. adding form_id to vault entries easily
2. avoid orphan vault entries in case of hard failures
3. keeps all saves together in the end of processing
Plus, we have not really "added" anything to HQ but just moved code around. So there is no additional load on core HQ to support this workflow, which would get removed if we fallback to saving vault entries upfront.

Reply:
I don't see a need for adding form ID to the vault entries, they don't have it now so why does saving them earlier require it? For 2 I think we just accept that there will be orphaned entries and leave it at that. Cleaning them up doesn't add much value.

Reply:
> I don't see a need for adding form ID to the vault entries

This is surely useful to be able to find the form for the vault entry which would be really difficult otherwise. I can see some value in that for debugging if not a direct feature implementation.

>  they don't have it now

we do have a column for form_id now.
https://github.com/dimagi/commcare-icds/pull/41/files#diff-b2b1b4b4d3111d1987ef6a08be0f9d42R200 
So we are saving them https://github.com/dimagi/commcare-icds/pull/41/files#diff-31f197c1e9dcb495b7aaa797a78d64b6R43